design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,398 @@
+  
+The way exceptions thrown by services are serialized right now for RESTful web API needs to be changed for inter-service endpoints.
+Exceptions are serialized without their type information provided - this info is crutial for services to communicate with each other so
+the type information will be added to the output.
+ 
+To make it easier for debuging and to perhaps later to be used for data consistency _X-Magento-Correlation-Id_ header will
+be added to requests to remote services with ID generated on BFF before a 1st request to a service. The ID will be
+forwarded between services to identify a single operation.  
+ 
+##### Retries mechanism
+The invoker implementation based on RESTful web APIs will offer default retries mechanism, but developers will be able to disable it
+by using _di.xml_ (see Invoker API and SPI [here](invoker.md)) for when they would want to delegate this responsibility
+to a service mash or a balancer. 
+The timeout and tries limit will be configured via di.xml (see the invoker API and SPI [here](invoker.md)).
+This default implementation may be used by developers who choose to avoid service mashes or for development environments.
+ 
+Regardless of which retries mechanism is applied when a timeout is configured a service may fail to deliver a result
+in time but will deliver it eventually, but the invoker (or a service mash) would've sent a duplicating request
+by that time. To avoid duplicating operations for such cases _X-Magento-Call-Id_ header
+will be added to inter-service requests to identify calls.
+It will be randomly generated. Then if a timeout reached while waiting for a service to respond the invoker
+(or a service mash) will send the same request
+again configured number of times with the same call ID. The service node receiving requests will store the call ID in cache as a key and
+assign an empty value to it. When the service node will have the service call result it will store it with the call ID. That way when a
+duplicate request comes in if the service node can't find the call ID in cache it will execute the requested service method as usual,
+if the cache has the call ID that would mean the service method is being executed and we'll have to wait until the original process
+finishes it and writes the response to cache, if both call ID and the result are in cache then we just return the result. That way
+we will avoid executing duplicate calls when retries mechanism is applied.
+
+#### Data consistency
+Mechanisms to preserve data consistency will described later in another document and can be built on
+top of the communications described in this one.
+ 
+    
+#### Service mashes and balancers
+Since we going to contact services via HTTP it is possible for configured service base URLs to actually lead to service mashes or
+balancers - that way the retries and lookup mechanism may be delegated.
+
+#### Deployment
```

Comment:
Probably the deployment topic should be described in a separate document as it much bigger than described in this proposal and at that moment we still don't have the final solution for services communication.